sources/packages/backend/libs/utilities/src/parallel-processing-utils.ts

```diff
@@ -23,14 +23,27 @@ export enum ParallelIntensity {
  * @param createPromise Implementation that calls processing function which
  * returns a promise.
  * @param inputs Input data from which the bulk processing is done.
- * @param maxParallelRequests Maximum number of parallel processes allowed to be executed at same time.
+ * @param options.
+ * - `maxParallelRequests` maximum number of parallel processes allowed
+ * to be executed at same time.
+ * - `progress` reports the current progress of the parallel processing.
+ * The `currentRecord` parameter reports the number of records processed (not the index).
+ * The method optionally allow to process async operations.
+ * - `partialResults` results from the last processed batch. Every time the promises are
+ * awaited this method is called. The method optionally allow to process async operations.
  * @returns resolved responses of promise.
  */
 export const processInParallel = async <P, I>(
   createPromise: (input: I) => Promise<P>,
   inputs: I[],
-  maxParallelRequests: ParallelIntensity = ParallelIntensity.Regular,
+  options?: {
+    maxParallelRequests?: ParallelIntensity;
+    progress?: (currentRecord: number) => Promise<void> | void;
+    partialResults?: (results: P[]) => Promise<void> | void;
```

Comment:
what about `processPartialResults` ?

Reply:
For `progress` it could be `updateProgress`.

Reply:
`processPartialResults` can indicate that a process will be executed with the records, which may be or may not be the case.
Yes, the current method consuming it is executing some processing, but the `processInParallel` is just communicating that some partial results are done.
Between `progress` and `updateProgress` I would stick with `progress`.
Please let me know if some of those are not suggestions.